src/utils/validation.ts

```diff
@@ -0,0 +1,65 @@
+import { providerErrors, rpcErrors } from '@metamask/rpc-errors';
+import type { Struct, StructError } from '@metamask/superstruct';
+import { validate } from '@metamask/superstruct';
+import type { Hex, JsonRpcRequest } from '@metamask/utils';
+
+export async function validateAndNormalizeKeyholder(
+  address: Hex,
+  req: JsonRpcRequest,
+  { getAccounts }: { getAccounts: (req: JsonRpcRequest) => Promise<string[]> },
+): Promise<Hex> {
+  if (
+    typeof address === 'string' &&
+    address.length > 0 &&
+    resemblesAddress(address)
+  ) {
+    // Ensure that an "unauthorized" error is thrown if the requester
+    // does not have the `eth_accounts` permission.
+    const accounts = await getAccounts(req);
+
+    const normalizedAccounts: string[] = accounts.map((_address) =>
+      _address.toLowerCase(),
+    );
+
+    const normalizedAddress = address.toLowerCase() as Hex;
+
+    if (normalizedAccounts.includes(normalizedAddress)) {
+      return normalizedAddress;
+    }
+
+    throw providerErrors.unauthorized();
+  }
+
+  throw rpcErrors.invalidParams({
+    message: `Invalid parameters: must provide an Ethereum address.`,
+  });
+}
+
+export function validateParams<ParamsType>(
+  value: unknown | ParamsType,
+  struct: Struct<ParamsType>,
+): value is ParamsType {
```

Comment:
seems like you might want the following return type instead:

```suggestion
): asserts value is ParamsType {
```

This would allow you to drop the return value in this helper and also the if statement guards in the handlers that call this

Reply:
Done.